lib/bolt/inventory.rb

```diff
@@ -243,5 +253,39 @@ def set_facts(tname, hash)
       end
     end
     private :set_facts
+
+    def add_node(grp, target, group_name, track = { 'all' => nil })
+      grp.groups.each do |g|
+        track[g.name] = { name: grp.name, data: grp }
+        add_node(g, target, group_name, track) if g.groups.any?
+        next unless g.name == group_name
+        # group name found, calculate data and insert new target to group
+        t_name = target.name
+        g.nodes[t_name] = { 'name' => t_name }.merge(target.options)
+        # inherit facts, vars, features from the group hierarchy, then merge with current level
+        data = inherit_data(track, g.name)
+        data[:facts].merge!(g.facts)
+        data[:vars].merge!(g.vars)
+        data[:features].concat(g.features)
+        set_facts(t_name, @target_facts[t_name] ? data[:facts].merge(@target_facts[t_name]) : data[:facts])
+        set_vars_from_hash(t_name, @target_vars[t_name] ? data[:vars].merge(@target_vars[t_name]) : data[:vars])
+        data[:features].each do |feature|
+          set_feature(target, feature)
+        end
+        return true
+      end
+    end
+    private :add_node
+
+    def inherit_data(track, name, acc = { facts: {}, vars: {}, features: [] })
+      if track[name]
```

Comment:
This is just checking for nil right? `name.nil?` seems easier to understand. perhaps `inherit_data(track, track[name][:name], acc) unless track[name][:name].nil?`